File: src/main/kotlin/com/cout970/magneticraft/api/util/ApiUtils.java
package com.cout970.magneticraft.api.util;

import net.minecraft.item.ItemStack;

/**
 * Created by cout970 on 16/06/2016.
 */
public class ApiUtils {

    private ApiUtils() {
    }

    public static boolean equalsIgnoreSize(ItemStack a, ItemStack b) {
        return a == b || !(a == null || b == null)
                         && (a.getItem().equals(b.getItem()))
                         && (!a.getItem().getHasSubtypes()
                             || (a.getMetadata() == b.getMetadata()))
                         && equals(a.getTagCompound(), b.getTagCompound());
    }

    public static boolean equals(Object a, Object b) {
        if (a == null) return b == null;
        return a.equals(b);
    }
}

File: src/main/kotlin/com/cout970/magneticraft/api/internal/ApiUtils.kt
package com.cout970.magneticraft.api.internal

import net.minecraft.item.ItemStack

/**
 * Created by cout970 on 16/06/2016.
 */
object ApiUtils {

    fun equalsIgnoreSize(a: ItemStack, b: ItemStack): Boolean {
        return a == b || !(a.isEmpty || b.isEmpty)
            && a.item == b.item
            && (!a.item.isDamageable || a.damage == b.damage)
            && a.tag == b.tag
    }

    fun equalsIgnoreSizeAndNBT(a: ItemStack, b: ItemStack): Boolean {
        return a == b || !(a.isEmpty || b.isEmpty)
            && a.item == b.item
            && (!a.item.isDamageable || a.damage == b.damage)
    }

    fun areEquivalent(a: ItemStack, b: ItemStack): Boolean {
        // TODO add OreDictionary functionality back
        return false
    }
}
